pkg/cosign/verify.go

```diff
@@ -986,52 +1004,44 @@ func VerifyBundle(ctx context.Context, sig oci.Signature, co *CheckOpts, rekorCl
 	return true, nil
 }
 
-func VerifyRFC3161Timestamp(ctx context.Context, sig oci.Signature, tsaCerts *x509.CertPool) (bool, error) {
+// VerifyRFC3161Timestamp verifies that the timestamp in untrustedSig is correctly signed, and if so,
+// returns the timestamp value.
+// It returns (nil, nil) if there is no timestamp; or (nil, err) if there is an invalid timestamp.
+func VerifyRFC3161Timestamp(ctx context.Context, sig oci.Signature, tsaCerts *x509.CertPool) (*timestamp.Timestamp, error) {
 	bundle, err := sig.RFC3161Timestamp()
 	if err != nil {
-		return false, err
+		return nil, err
 	} else if bundle == nil {
-		return false, nil
+		return nil, nil
 	}
 
 	b64Sig, err := sig.Base64Signature()
 	if err != nil {
-		return false, fmt.Errorf("reading base64signature: %w", err)
-	}
-
-	cert, err := sig.Cert()
-	if err != nil {
-		return false, err
+		return nil, fmt.Errorf("reading base64signature: %w", err)
 	}
 
 	verifiedBytes := []byte(b64Sig)
 	if len(b64Sig) == 0 {
 		// For attestations, the Base64Signature is not set, therefore we rely on the signed payload
 		signedPayload, err := sig.Payload()
 		if err != nil {
-			return false, fmt.Errorf("reading the payload: %w", err)
+			return nil, fmt.Errorf("reading the payload: %w", err)
 		}
 		verifiedBytes = signedPayload
 	}
 
 	err = tsaverification.VerifyTimestampResponse(bundle.SignedRFC3161Timestamp, bytes.NewReader(verifiedBytes), tsaCerts)
 	if err != nil {
-		return false, fmt.Errorf("unable to verify TimestampResponse: %w", err)
+		return nil, fmt.Errorf("unable to verify TimestampResponse: %w", err)
 	}
+	acceptedTimestamp := bundle
 
-	if cert != nil {
-		ts, err := timestamp.ParseResponse(bundle.SignedRFC3161Timestamp)
-		if err != nil {
-			return false, fmt.Errorf("error parsing response into timestamp: %w", err)
-		}
-		// Verify the cert against the integrated time.
-		// Note that if the caller requires the certificate to be present, it has to ensure that itself.
-		if err := CheckExpiry(cert, ts.Time); err != nil {
-			return false, fmt.Errorf("checking expiry on cert: %w", err)
-		}
+	// FIXME: tsaverification.VerifyTimestampResponse has done this parsing; we shouldn’t parse again.
```

Comment:
please file feedback in sigstore/timestamp-authority for updating the verifier

Reply:
https://github.com/sigstore/timestamp-authority/issues/169